File: 2022/clojure/src/advent/x1.clj
(ns advent.x1
  (:gen-class))

(require '[clojure.string :as str])

(def parse
  (->> (str/split (slurp "../inputs/1") #"\n\n")
       (map #(str/split %1 #"\n"))
       (map #(map (fn [x] (Integer/parseInt x)) %))
       (map #(apply + %))
       ))

(def part-one (reduce max parse))
(def part-two (->> parse
               sort
               reverse
               (take 3)
               (apply +)))

(defn -main [& args]
  (println part-one)
  (println part-two))

File: 2015/haskell/app/X1.hs
import Lude

main :: IO ()
main = do
  input <- readFile "inputs/1"

  -- part one
  let parse '(' = 1
      parse ')' = -1
      parse  _  = 0
  print (input & map parse & sum)

  -- part two
  print (input & map parse & scanl1 (+) & takeWhile (>= 0) & length & (+ 1))
